File: Medium/[2130] Maximum Twin Sum of a Linked List/Python3 Solution/[2130] Maximum Twin Sum of a Linked List.py
"""
Accepted
2130 [Medium]
Runtime: 99 ms, faster than 11.12% of Python3 online submissions for Maximum Twin Sum of a Linked List.
Memory Usage: 66.31 MB, less than 5.03% of Python3 online submissions for Maximum Twin Sum of a Linked List.
"""
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        # Recursive Solution
        if head == None or head.next == None:
            return head
        p = self.reverseList(head.next)
        head.next.next = head
        head.next = None
        return p

    def pairSum(self, head: Optional[ListNode]) -> int:
        # Count the number of nodes
        n = 0
        ptr = head
        while ptr:
            n += 1
            ptr = ptr.next

        ptr = head
        for i in range(n // 2 - 1):
            ptr = ptr.next

        # Reverse the second half
        ptr.next = self.reverseList(ptr.next)
        ptr = ptr.next

        res = 0
        cur = head
        while ptr:
            res = max(res, cur.val + ptr.val)
            cur = cur.next
            ptr = ptr.next
        return res

File: Medium/[2130] Maximum Twin Sum of a Linked List/Java Solution/[2130] Maximum Twin Sum of a Linked List.java
//Accepted
//2130 [Medium]
//Runtime: 11 ms, faster than 12.92% of Java online submissions for Maximum Twin Sum of a Linked List.
//Memory Usage: 112.71 MB, less than 19.99% of Java online submissions for Maximum Twin Sum of a Linked List.
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public int pairSum(ListNode head) {
        ListNode ptr=head;
        int sum,max=0,i,count=0;
        while(ptr!=null) {
            count++;
            ptr=ptr.next;
        }
        int n[]=new int[count];
        ptr=head;
        for(i=0;i<count;i++) {
            if(ptr!=null) {
                n[i]=ptr.val;
                ptr=ptr.next;
            }
            else
                break;
        }
        for(i=0;i<n.length/2;i++) {
            sum=n[i]+n[n.length-1-i];
            if(max<sum)
                max=sum;
        }
        return max;
    }
}